tests/test_websocket_bootstrap.c

```diff
@@ -0,0 +1,635 @@
+    AWS_FATAL_ASSERT(!s_tester.http_stream_release_called);
+    return s_mock_http_connection;
+}
+
+static int s_mock_http_stream_get_incoming_response_status(const struct aws_http_stream *stream, int *out_status) {
+    AWS_FATAL_ASSERT(stream == s_mock_stream);
+    AWS_FATAL_ASSERT(!s_tester.http_connection_release_called);
+    AWS_FATAL_ASSERT(!s_tester.http_stream_release_called);
+
+    if (s_tester.fail_at_step == BOOT_STEP_VALIDATE_RESPONSE_STATUS) {
+        *out_status = 403;
+    } else {
+        *out_status = 101;
+    }
+    return AWS_OP_SUCCESS;
+}
+
+static struct aws_websocket *s_mock_websocket_handler_new(const struct aws_websocket_handler_options *options) {
+    AWS_FATAL_ASSERT(options);
+    AWS_FATAL_ASSERT(!s_tester.http_connection_release_called);
+    AWS_FATAL_ASSERT(!s_tester.websocket_new_called_successfully); /* ensure we're only called once */
+
+    if (s_tester.fail_at_step == BOOT_STEP_WEBSOCKET_NEW) {
+        aws_raise_error(BOOT_STEP_WEBSOCKET_NEW);
+        return NULL;
+    }
+
+    s_tester.websocket_new_called_successfully = true;
+    return s_mock_websocket;
+}
+
+static void s_on_websocket_setup(
+    struct aws_websocket *websocket,
+    int error_code,
+    int handshake_response_status,
+    const struct aws_http_header *handshake_response_header_array,
+    size_t num_handshake_response_headers,
+    void *user_data) {
+
+    if (error_code) {
+        AWS_FATAL_ASSERT(!websocket);
+    } else {
+        AWS_FATAL_ASSERT(websocket == s_mock_websocket);
+
+        /* Check that headers passed by mock response carry through. */
+        AWS_FATAL_ASSERT(s_headers_eq(
+            s_tester.handshake_response_headers,
+            s_tester.num_handshake_response_headers,
+            handshake_response_header_array,
+            num_handshake_response_headers));
+
+        AWS_FATAL_ASSERT(handshake_response_status == 101);
+    }
+
+    AWS_FATAL_ASSERT(user_data == &s_tester);
+
+    s_tester.websocket_setup_invoked = true;
+    s_tester.websocket_setup_error_code = error_code;
+}
+
+static void s_on_websocket_shutdown(struct aws_websocket *websocket, int error_code, void *user_data) {
+    AWS_FATAL_ASSERT(websocket == s_mock_websocket);
+    AWS_FATAL_ASSERT(user_data == &s_tester);
+
+    s_tester.websocket_shutdown_invoked = true;
+    s_tester.websocket_shutdown_error_code = error_code;
+}
+
+/* Calls aws_websocket_client_connect(), and drives the async call to its conclusions.
+ * Reports the reason for the failure via `out_error_code`. */
+static int s_drive_websocket_connect(int *out_error_code) {
+    bool websocket_connect_called_successfully = false;
+    bool http_connect_setup_reported_success = false;
+
+    /* Call websocket_connect() */
+    struct aws_byte_cursor uri_cursor = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("server.example.com");
+    struct aws_uri uri;
+    int err = aws_uri_init_parse(&uri, s_tester.alloc, &uri_cursor);
+    if (err) {
```

Comment:
This feels like it should be an assert.  If there's a failure you'll only detect it "downstream" with asserts on later expected state.

Reply:
This only fails in my out-of-memory test when 0 allocations are allowed. There's a `goto finishing_checks` in the next line